aten/src/ATen/Context.cpp

```diff
@@ -78,6 +78,29 @@ void Context::alertNotDeterministic(c10::string_view const& caller) {
   }
 }
 
+void Context::alertCuBLASConfigNotDeterministic() {
+  if (deterministic() && hasCUDART() && (versionCUDART() >= 10020)) {
+    // If using CUDA 10.2 or greater, need to make sure CuBLAS workspace config
+    // is set to deterministic setting
+    static const char config_var_name[] = "CUBLAS_WORKSPACE_CONFIG";
+    static const char config0[] = ":4096:8";
+    static const char config1[] = ":16:8";
+    char* workspace_config = std::getenv(config_var_name);
+    bool workspace_config_is_deterministic = (workspace_config != nullptr) && (
+      (strcmp(workspace_config, config0) == 0) || (strcmp(workspace_config, config1) == 0)
+    );
```

Comment:
In general, we should probably assume that the environment variable stays fixed throughout process lifetime, and test it only once (rather what you are doing now, which is do a string comparison EVERY time someone calls in to alertCuBLASConfigNotDeterministic). I also don't know if you verified that cublas even respects the environment variable if you set it midway through a process: I suspect it does not.

Reply:
I did check, and cuBLAS does not respect it if you change the variable midway, it only checks the variable once. I'll change this to only check the variable once and cache the result.

Reply:
Actually, `alertCuBLASConfigNotDeterministic` only checks once and caches the result, that would make it hard to unit test it within one process.

Right now, the `checkAlertCuBLASConfigNondeterministic` decorator that I wrote calls the affected CuBLAS functions several times, changing the environment variable in between. Depending on the value it sets, it expects the function to either work or throw an error. Note that this decorator doesn't care if the behavior of the function is actually deterministic or not, it's just an API test, so it's alright that changing the variable at run time doesn't really change CuBLAS's behavior.

If I changed `alertCuBLASConfigNotDeterministic` to cache the value, then I don't think we can easily write a proper test for the affected functions. Unless there's something I'm not thinking of.

Reply:
We have some tests which spawn new Python processes to test for situations where you can't easily test in one process, it seems applicable to this case.

Reply:
Ah ok, I'll do that.

Reply:
I'm not 100% sure how to do this yet. If possible, I'd like to keep the process spawning all contained in the `checkAlertCuBLASConfigNotDeterministic` decorator. So I tried to do this inside the wrapper:

```
torch.multiprocessing.spawn(
    expectCuBLASConfigNotDeterministicError,  // this is just a function that will run `fn` and check for the error
    args=(fn, device, args, kwargs),
    join=True
```

But when I run this, I get `AttributeError: Can't pickle local object 'add_test.<locals>.do_test'`, referring to the `fn`. I'm not sure how to get around this.

Reply:
I think the explanation is that the `do_test` function in test/test_autograd.py is not at the top level of the module. So enabling it for pickling would probably require significant changes. Maybe it would be better to just avoid trying to use `common_methods_invocations.py` for these test cases, in favor of a more specialized test in `test_torch.py`.

Reply:
Alright, I changed `alertCuBLASConfigNontDeterministic` to check the env var only once per process and wrote a test that spawns processes for each test case.